app/controllers/concerns/foreman/controller/discovered_extensions.rb

```diff
@@ -36,7 +37,17 @@ def perform_auto_provision host, rule
     host.hostgroup_id = rule.hostgroup_id
     host.comment = "Auto-discovered and provisioned via rule '#{rule.name}'"
     host.discovery_rule = rule
-    host.save # save! does not work here
+    if host.save #save! does not work here
+      delete_discovery_attribute_set(host.id)
+    end
+  end
+
+  # discovery_attribute_set should be deleted only after a host has been successfully provisioned
+  # (after discovered host becomes managed host)
+  # after this happens the host is a manged host so non of the callbacks in discovered hosts are relevant anymore
+  # for this reason the orphaned discovery_attribute_set needs to be searched for and deleted
+  def delete_discovery_attribute_set(host_id)
+    DiscoveryAttributeSet.delete_all(:host_id => host_id)
```

Comment:
nitpick: would it be better to pass the discovery attribute set id instead of the host id?